bin/database.py

```
@@ -2426,32 +2367,42 @@ def upgrade_schema(force_from = None):
         logging.error('The stored db schema version of %s is incompatible with required version %s',
                        str(db_version), CURRENT_DATABASE_VERSION)
         sys.exit(43)
-    elif db_version == CURRENT_DATABASE_VERSION:
+    elif db_version == CURRENT_DATABASE_VERSION and not available_fixes:
         logging.error('Database already up to date.')
         sys.exit(43)
 
+    update_doc = {}
+
     try:
         while db_version < CURRENT_DATABASE_VERSION:
+            # Apply fixes before performing the next schema update
+            apply_available_fixes(db_version, applied_fixes, update_doc)
+
             db_version += 1
             upgrade_script = 'upgrade_to_'+str(db_version)
             logging.info('Upgrading to version {} ...'.format(db_version))
             globals()[upgrade_script]()
             logging.info('Upgrade to version {} complete.'.format(db_version))
+
+        # Last round of fixes for the current db version
+        apply_available_fixes(db_version, applied_fixes, update_doc)
     except KeyError as e:
         logging.exception('Attempted to upgrade using script that does not exist: {}'.format(e))
         sys.exit(1)
     except Exception as e:
         logging.exception('Incremental upgrade of db failed')
         sys.exit(1)
     else:
-        config.db.singletons.update_one({'_id': 'version'}, {'$set': {'database': CURRENT_DATABASE_VERSION}})
+        update_doc['database'] = CURRENT_DATABASE_VERSION
+        config.db.singletons.update_one({'_id': 'version'}, {'$set': update_doc})
         sys.exit(0)
 
 if __name__ == '__main__':
     try:
         parser = argparse.ArgumentParser()
         parser.add_argument("function", help="function to be called from database.py")
         parser.add_argument("-f", "--force_from", help="force database to upgrade from previous version", type=int)
+        parser.add_argument("-F", "--fix-id", help="ID of fix to run if applying a fix")
         args = parser.parse_args()
```

It's a bit confusing to have a `-f` and `-F` that do mostly unrelated things but not the end of the world.

---

Thinking about it a little longer .. if you got it wrong it would be a fast failure. Only worry would be if it was a cryptic/harmful/silent failure. Let's just keep it as is.